include/exadg/incompressible_navier_stokes/user_interface/parameters.cpp

```diff
@@ -559,23 +636,61 @@ bool
 Parameters::viscous_problem() const
 {
   return (equation_type == EquationType::Stokes || equation_type == EquationType::NavierStokes ||
-          use_turbulence_model == true);
+          viscosity_model_data.use_turbulence_model == true);
+}
+
+bool
+Parameters::viscous_term_is_linear() const
+{
+  if(viscosity_is_variable() && viscosity_model_data.treatment_of_nonlinear_viscosity ==
+                                  TreatmentOfNonlinearViscosity::Implicit)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool
+Parameters::viscous_term_is_nonlinear() const
+{
+  return (viscous_term_is_linear() == false);
+}
+
+bool
+Parameters::viscosity_is_variable() const
+{
+  return viscosity_model_data.use_turbulence_model ||
+         viscosity_model_data.use_generalized_newtonian_model;
 }
 
 bool
 Parameters::nonlinear_problem_has_to_be_solved() const
 {
-  return convective_problem() &&
-         (solver_type == SolverType::Steady ||
-          (solver_type == SolverType::Unsteady &&
-           treatment_of_convective_term == TreatmentOfConvectiveTerm::Implicit));
+  if(convective_problem())
+  {
+    if(solver_type == SolverType::Steady ||
+       (solver_type == SolverType::Unsteady &&
+        (treatment_of_convective_term == TreatmentOfConvectiveTerm::Implicit)))
+    {
+      return true;
+    }
+  }
+
+  if(viscous_problem() && viscous_term_is_nonlinear() &&
+     viscosity_model_data.treatment_of_nonlinear_viscosity ==
+       TreatmentOfNonlinearViscosity::Implicit)
+  {
+    return true;
+  }
+
+  return false;
```

Comment:
```suggestion
  if(convective_problem() and convective_term_is_solved_implicitly()
        return true;
  else
      return (viscous_problem() && viscous_term_is_nonlinear() &&
     viscosity_model_data.treatment_of_nonlinear_viscosity ==
       TreatmentOfNonlinearViscosity::Implicit);
```

including @nfehn 's suggestion and  simplifying a bit.

Reply:
Here, I am of a different opinion. I find multiple if-statements with a final `return false` as suggested by @richardschu  easier to read/understand than the suggestion above.